### PR TITLE
[DEV-1718] Avoid unnecessary recalculation and upserting of on-demand tiles for historical features

### PR DESCRIPTION
**Pull request by YS-L**
## Description

This updates the tile cache compute logic to avoid unnecessary recalculation and upserting of on-demand tiles for historical features. Before this change, on-demand tiles for historical features are always computed from the beginning of time. That is non-ideal as both the recalculation and upserting can be expensive.

Using the [`CustomerInventoryConsistency_6w90d`](https://github.com/featurebyte/featurebyte-notebooks/blob/0bff4241f46780f7498a43ceeaf2fdd758eb9630/notebooks/features/grocery/features_set_5.py#L72-L75) feature for testing, we can simulate a scenario where a new observation set contains more recent point in time values and causes tile cache to be outdated when computing historical features. A comparison before vs after:

* Runtime for on-demand tile compute and upsert query: 31 minutes (before) vs 53s (after).
* Number of rows merged into the tile table: 6B rows (before) vs 6M rows (after).

More details on the tests can be found here: https://github.com/featurebyte/featurebyte-notebooks/pull/3

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

**Review by yanhuisg — APPROVED** [review]
LGTM

File: featurebyte/tile/tile_cache.py [hunk 1]
```diff
@@ -311,54 +314,54 @@ async def _filter_tile_ids_with_tracker(self, tile_ids: list[str]) -> list[str]:
...
+        agg_ids_with_tracker: list[str],
+        agg_ids_no_tracker: list[str],
         request_id: str,
         request_table_name: str,
     ) -> None:
         """Register a temp table from which we can query whether each (POINT_IN_TIME, ENTITY_ID,
-        TILE_ID) pair has updated tile cache: a null value in this table indicates that the pair has
-        outdated tile cache. A non-null value refers to the valid last tile start date registered in
-        the data warehouse's tracking table for that pair.
+        TILE_ID) pair has updated tile cache:
```
Comment (gee-senbong) [inline featurebyte/tile/tile_cache.py:336]: nit: `each (POINT_IN_TIME, ENTITY_ID, TILE_ID) triple`

File: featurebyte/tile/tile_cache.py [hunk 2]
```diff
@@ -384,24 +383,18 @@ async def _register_working_table(
...
-                )
-            )
+            # Note: join_conditions is empty list if there is no entity column. In this case, there
+            # is only one row in the tracking table and the join condition can be omitted.
             table_expr = table_expr.join(
                 tracker_table_name,
                 join_type="left",
                 join_alias=table_alias,
-                on=expressions.and_(*join_conditions),
+                on=expressions.and_(*join_conditions) if join_conditions else None,
```
Comment (gee-senbong) [inline featurebyte/tile/tile_cache.py:392]: Curious whether we have existing test case cover the case where `join_conditions` is empty list?

**Review by gee-senbong — APPROVED** [review]
LTGM. Left some comments to consider.

**Review by kchua78 — APPROVED** [review]
:+1:

File: featurebyte/tile/tile_cache.py (hunk 2 above)
Comment (YS-L) [reply featurebyte/tile/tile_cache.py:392]: Yes, this test covers that scenario: https://github.com/featurebyte/featurebyte/blob/e9b628e40c511383e70dec55cd165341658d291d/tests/integration/api/test_change_view_operations.py#L69

The test was failing without the special handling here.

**Pull request merged**